Exact tests comparing to QD

JanosJiri opened the pull request:
In this pull request, I added two tests comparing results from the code with results used in the publication and compared to QD. This test will be essential before we can start moving some important parts of the code inside `InitialConditions`. At the moment, the PDAW reference was not compared to QD, but I'm working on it. The PDA was compared to QD, see attached figure. The tests will need refactoring.
![plot](https://github.com/user-attachments/assets/e5a41652-3ffa-4193-a488-d3bb10951a8a)

JanosJiri commented:
I added new reference data for the PDAW test. This set was benchmarked against QD. Therefore, we have now tests for both PDA and PDAW that check it against QD. However, the numerics again failed, this time for PDAW (I added some preliminary rounding for PDA which made it pass the test). The test worked on my computer where I generated the data. I think the numerical difference `assert np.float64(5.033296566754934e-22) == np.float64(5.033296466962208e-22)` is negligible but I don't know how to correctly compare numbers within numerical accuracy. I hope for help from @danielhollas once you have a bit of time. 
![plot](https://github.com/user-attachments/assets/a574c6a8-0e7b-42bf-bba6-07f6c6c4179e)

danielhollas commented:
This is great, thanks for doing this! Definitely agree we need more "end-to-end" tests. 

Comparing real numbers is always a PITA, I have always struggled with that in ABIN tests. I don't actually have experience with handling this problem in Python so we'll need to figure it out.

For now it would be good to understand where the differences come from. Can you post what Python and numpy versions do you have locally? (`pip show numpy` should show you the package version).

> This set was benchmarked against QD. Therefore, we have now tests for both PDA and PDAW that check it against QD.

Awesome! :+1:

danielhollas commented on pyproject.toml:
```diff
@@ -1,6 +1,10 @@
 version = "0.1.0"
+authors = [
+  { name="Jiri Janos", email="jiri.janos98@gmail.com" },
+  { name="Daniel Hollas", email="danekhollas@gmail.com" },
```
:heart:

danielhollas commented on tests/test_methods_nai.py:
```diff
@@ -0,0 +1,58 @@
+"""This test aims to compare to previous PDA and PDAW results that were published and compared to exact QD.
```
Could you tell me which version of the code specifically (commit hash ideally) did you use to generate the reference data?

JanosJiri replied:
I used the last version on main (commit [629117a](https://github.com/PHOTOX/promdens/commit/629117adc7b0b5d24627375cb221ccea13b074df)) to generate the reference data. I created the data today and compared with QD.

danielhollas replied:
Okay cool. Would be also good to verify that you get the same numbers with the older versions of the code before I started to mess with it. :smile:

JanosJiri replied:
I was doing the same test before we put it on GitHub and it passed. I'm not comparing the numbers but I always compared with QD and also tested that it wouldn't work if you mess up with the results. 
Just to make sure, I just tried two other pulses published (Gaussian and Lorentzian) and I still get the same as QD (one results attached). Thus, I believe this version is stable and working.

<img width="389" alt="Screenshot 2024-09-13 at 18 30 36" src="https://github.com/user-attachments/assets/d7f786ca-21e6-4574-98d3-1274b3e2f010">

danielhollas commented on tests/test_methods_nai.py:
```diff
@@ -0,0 +1,58 @@
+from promdens.promdens import InitialConditions, LaserPulse
+
+
+def test_pdaw_nai(make_pulse):
```
```suggestion
def test_pdaw_nai():
```

`make_pulse` fixture is unused in this test, and also the test below.

danielhollas commented on tests/test_methods_nai.py:
```diff
@@ -0,0 +1,58 @@
+    ics.windowing()
+
+    # the following command was used to generate the reference
+    # promdens -m pdaw -n 1000 -w 0.13520905 -f 20 test_data_nai.dat
```
Thanks for adding the comments!

danielhollas commented on tests/test_methods_nai.py:
```diff
@@ -0,0 +1,58 @@
+
+    ics.calc_field(pulse=pulse)
+
+    ics.sample_initial_conditions(nsamples_ic=500, neg_handling='error', preselect=True, seed=123456789)
```
This takes a fairly long time, I think for the purposes of the test we want to lower the `nsamples_ic` significantly.

JanosJiri replied:
I know it is a longer test but I made it longer on purpose. With less than 500 samples, I don't have enough data to make sure I match QD. I felt like 10 seconds was not too much but I can make it shorter if you think that large test is unnecessary.

danielhollas replied:
I think we should decouple the validation with QD from the unit tests defined in the `tests/` directory, as they serve slightly different roles. I'll write more about that later.

JanosJiri replied:
I will change it to 10 samples to make it faster and later we can discuss if we also want to have some validation.

danielhollas commented on tests/test_methods_nai.py:
```diff
@@ -0,0 +1,58 @@
+def test_pdaw_nai(make_pulse):
+    ics = InitialConditions(nsamples=1000, nstates=1, input_type='file')
+
+    # todo: this file path works only if 'pytest' is launched in the main folder (the same for all later)
```
You can do

```python
reference_file = Path(__file__).parent.absolute() / 'NaI_reference/test_pdaw_reference.dat'
```

danielhollas replied:
You'll also have to add `from pathlib import Path` at the top

danielhollas commented on tests/test_methods_nai.py:
```diff
@@ -0,0 +1,58 @@
+        assert weights[i] == reference[i]
+
+
+def test_pda_nai(make_pulse):
```
Please use the builtin `[tmp_path](https://docs.pytest.org/en/stable/how-to/tmp_path.html#the-tmp-path-fixture) fixture so that we do not produce the `pda.dat` file inside the repo (same for the other tests).

danielhollas commented on tests/test_methods_nai.py:
```diff
@@ -0,0 +1,58 @@
+    assert len(weights) == len(weights)
+    # comparing all weights
+    for i in range(len(weights)):
+        assert weights[i] == reference[i]
```
This will make the output of a failed test little more informative
```suggestion
        assert weights[i] == reference[i], f"pdaw weight[{i}] does not match: delta={weights[i]-reference[i]}"
```

danielhollas reviewed (changes requested):
Some suggestions (not trying to solve the numerical issue yet)

JanosJiri commented:
> For now it would be good to understand where the differences come from. Can you post what Python and numpy versions do you have locally? (`pip show numpy` should show you the package version).

I used Python 3.12.1 and numpy 1.26.2 to create the reference data.

danielhollas commented:
I tried with Python 3.12.5 and numpy 1.26.2 but still am seeing differences locally. That's unfortunate, as it likely means that the results will be sensitive on the CPU.

It looks like [numpy.testing](https://numpy.org/doc/stable/reference/routines.testing.html) module might be useful here, perhaps  [assert_allclose](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose)

JanosJiri commented:
Thanks for having a look at the pull request @danielhollas ! I made five changes:

- I added absolute paths to all files read by the test (as you advised).
- I created tmp_path folders and moved output files of promdens there.
- I modified the preliminary numeric test. Now I compare if the difference between two numbers is bigger than some threshold. The threshold was based on my own comparison of results generated on the same computer by numpy 1.26 and 2.1
- I changed the reference data to values generated with numpy 2.1
- I changed the PDA test to only 10 ICs. I again compared the code with some QD simulations with different pulses. I believe this version is correct and stable. Thus, the reference data generated should be good ones.

Let me know if you have any further suggestions and if you find some smart way to deal with the numeric tests. But with this, I will feel safer if we start moving functions around the code as we test that the physics calculated still work.

danielhollas commented on tests/test_methods_nai.py:
```diff
@@ -0,0 +1,71 @@
+    ics.windowing()
+
+    # moving output to a temporary path dedicated to this test
+    tmp_folder = tmp_path / "pdaw"
```
`tmp_path` and is unique for this test so you can just use it instead of creating a subfolder. (same for the other test)

danielhollas commented on tests/test_methods_nai.py:
```diff
@@ -0,0 +1,71 @@
+    # moving output to a temporary path dedicated to this test
+    tmp_folder = tmp_path / "pdaw"
+    tmp_folder.mkdir()
+    Path('pdaw.dat').rename(tmp_folder / 'pdaw.dat')
```
the Path().rename() method unfortunately does not work if the original and target paths are not on the same filesystem. On my laptop I am getting this error

```
OSError: [Errno 18] Invalid cross-device link: 
'pdaw.dat' -> '/tmp/pytest-of-hollas/pytest-8/test_pdaw_nai0/pdaw/pdaw.dat'
```

Instead I think the file shoud be created in `tmp_path` in the first place. I'd add an extra argument to `windowing` function, perhaps names `output_fname` or something like that so you can pass the path when calling the function. And same for `sample_initial_conditions`. The arguments can have default values so you don't have to change the code in the main script and only use it here in tests. WDYT?

```
    def windowing(self, output_fname='pdaw.dat'):
```

danielhollas reviewed (changes requested):
@JanosJiri perfect, thank you for the updates! I have one more thing about the tests, but otherwise looks good. 

For the validation with QD, we could create a separate folder called `validation`, move the `EnvelopeAnalysis` into it, and have another QD_comparison folder, which would contain the whole input and output files, for both pda and pdaw and could also include the pictures you posted in this PR. That way it does not matter that the test themselves only test a subset of data, and we'll have a record of the full validation that can be recreated if needed. (perhaps this should be a separate PR).

danielhollas commented on tests/NaI_reference/test_data_nai.dat:
With the comment about the separate validation folder, in this test folder I would only include the ~10 ICs that we test. (that way the `test_pdaw_reference.dat` wouldn't be that big, and would have less chance to have numerical difference in the future.

JanosJiri commented:
Hi @danielhollas , I processed all the comments from the previous review. We now test only 10 geometries to make the test faster and numerically more stable. The output in the test is also directly stored in the `tmp_path`. I also created a validation folder but saved it for another pull request which is coming soon. Hopefully done for this test. Thanks for all the reviews!

danielhollas reviewed (approved):
Perfect, great work, thanks for bearing with me! :bowing_man: I verified that the tests also pass for me locally.
